<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<link rel="stylesheet" media="all" href="/site_media/resources/bootstrap/css/bootstrap.min.css" />
<link href="/site_media/resources/kendo/styles/kendo.common.min.css" rel="stylesheet" />
<link href="/site_media/resources/less/kendo.tb.min.css" rel="stylesheet" />
<link type="text/css" rel="stylesheet" href="/site_media/resources/styles/tb-layout.css" />
<link type="text/css" rel="stylesheet" href="/site_media/resources/styles/tb-styles.min.css" />

<link rel="stylesheet" type="text/css" href="/site_media/stylesheet.css"/>
<link rel="stylesheet" type="text/css" href="/site_media/resources/styles/print.css" media="print" />
<link rel="stylesheet" type="text/css" href="/site_media/resources/styles/report.css" media="screen" />

<link rel="stylesheet" type="text/css" href="lifechart/tabbar.css"/>

<script type="text/javascript" src="/site_media/resources/jquery/jquery-1.8.2.min.js"></script>
<script type="text/javascript" src="/site_media/resources/scripts/kendo.custom.min.js"></script>

{% load humanize %}

{% if autorefresh %}
<META HTTP-EQUIV="refresh" CONTENT="15">
{% endif %}

<style type="text/css">
  body {background:white}
  .help {cursor:help; border-bottom: 1px dotted #A9A9A9}
</style>

</head>
<body>

<title>RNASeq Analysis Report</title>
<body>

<div class="container-fluid">

<h1><center>RNASeq Analysis Report</center></h1>
<h3><center>{{run_name}}<center></h3>
<h2><center>Barcode Summary</center></h2>

<h4>
<b>Reference:</b> {{ref_name}}<br/>
</h4>

<script type="text/javascript">
  function numberWithCommas(x) {
    var parts = x.toString().split(".");
    parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ",");
    return parts.join(".");
  }
  function numberToPercent(x) {
    if( !x ) return "NA";
    if( isNaN(x) ) return x;
    x *= 100;
    return x.toFixed(2)+"%";
  }
</script>

<!-- Capture django variable in to javascript -->
<script>
  var barcodes_json = {{barcode_results|safe}};
</script>

<script type="text/javascript">
$(document).ready(function() {
  if (typeof barcodes_json !== 'undefined'){
    $("#barcodes").kendoGrid({
      height: 'auto',
      groupable: false,
      scrollable: false,
      selectable: false,
      sortable: { mode: "multiple", allowUnsort: true },
      pageable : { pageSizes:[5,10,20,50,100,1000] },
      dataSource: {
        data: barcodes_json,
        schema: { model: { fields: {
          barcode_name: {type:"string"},
          sample: {type:"string"},
          total_reads: {type:"number"},
          aligned_reads: {type:"number"},
          pc_aliged_reads: {type:"number"},
          mean_read_length: {type:"number"},
          genes_detected: {type:"number"},
          isoforms_detected: {type:"number"}
        } } },
        pageSize: 20
      },
      columns: [
        {field:"barcode_name"},
        {field:"sample"},
        {field:"total_reads"},
        {field:"aligned_reads"},
        {field:"pc_aliged_reads"},
        {field:"mean_read_length"},
        {field:"genes_detected"},
        {field:"isoforms_detected"}
      ],
      rowTemplate: kendo.template($("#barcodesRowTemplate").html())
    });
  }
});
</script>

<div>
  <div class="spacer clear"></div>
  <table id="barcodes" style="width:100%">
    <thead>
    <tr>
      <th title="Barcode name and link to detailed report.">Barcode Name</th>
      <th title="Sample name.">Sample</th>
      <th title="Total number of reads.">Total Reads</th>
      <th title="Total number of reads aligned to the reference.">Aligned Reads</th>
      <th title="Percentage of total reads aligned to the reference.">Percent Aligned</th>
      <th title="Average aligned read length.">Mean Read Length</th>
      <th title="Number of annotated genes with at least 10 reads.">Genes Detected</th>
      <th title="Number of gene transcript isoforms annotated.">Isoforms Detected</th>
    </tr>
    </thead>
    <script id="barcodesRowTemplate" type="text/x-kendo-tmpl">
      <tr>
        <td> #= barcode_details #</td>
        <td> #= (barcode_name) ? sample : "None" #</td>
        <td> #=numberWithCommas(total_reads)# </td>
        <td> #=numberWithCommas(aligned_reads)# </td>
        <td> #=numberToPercent(pc_aliged_reads)# </td>
        <td> #=mean_read_length# </td>
        <td> #=numberWithCommas(genes_detected)# </td>
        <td> #=numberWithCommas(isoforms_detected)# </td>
      </tr>
    </script>
  </table>
</div>

{% if num_barcodes_filtered > 0 %}
<h3><span title="Reads for some barcodes were ignored as having less than {{barcode_filter}}% of the file size of the largest barcode aligned reads file. To include these barcodes in the analysis, re-run with the Filter Barcodes option unchecked." style="cursor:help;color:red">WARNING: {{num_barcodes_filtered}} barcodes were discounted from analysis as having relatively too few reads.</span></h3>
{% endif %}

<script type="text/javascript">
  $(document).ready(function() {
    $('.tabs .tab-title a').on('click', function(e)  {
      var currentAttrValue = $(this).attr('href');
      $('.tabs-content '+currentAttrValue).show().siblings().hide();
      $(this).parent('li').addClass('active').siblings().removeClass('active');
      e.preventDefault();
  });
});
</script>

<script type="text/javascript">
  var img_width = 0, img_height = 0 ;
  var img = new Image();
  img.onload = function() {
    img_width = this.width;
    img_height = this.height;
    var maxWidth = screen.availWidth-10;
    var maxHeight = screen.availHeight-10;
    if( img_width > maxWidth ) img_width = maxWidth;
    if( img_height > maxHeight ) img_height = maxHeight;
  }
  img.src = $("#paircor_img").attr("src");
  $(".unscaled_image_window").click(function() {
    window.open(this.href,"_blank",'height='+img_height+',width='+img_width+',titlebar=0,toolbar=0,location=0,status=0'); 
    return false;
  });
</script>

{% if bctable %}
<ul class="tabs" data-tab>
  <li class="tab-title active"><a href="#QAPlots">Distribution Plots</a></li>
{% if rvalheatmap %}
  <li class="tab-title"><a href="#CorHeatmap">Correlation Heatmap</a></li>
{% endif %}
{% if cpairsplot %}
  <li class="tab-title"><a href="#CorPlot">Correlation Plot</a></li>
{% endif %}
{% if geneheatmap %}
  <li class="tab-title"><a href="#GeneHeatmap">Gene Heatmap</a></li>
{% endif %}
{% if isofheatmap %}
  <li class="tab-title"><a href="#IsoformHeatmap">Isoform Heatmap</a></li>
{% endif %}
</ul>

<div class="tabs-content">
  <div class="content active" id="QAPlots">
    <table style="margin-left:auto;margin-right:auto;"><tr>
    {% if mapreadsplot %}
      <td><a href="{{mapreadsplot}}" title="A graphical summary of the numbers of mapped and unmapped across bacodes, as reported in the Bacode Summary table."><img style="height:auto" src="{{mapreadsplot}}" alt="Reads Alignment Summary"/></a></td>
    {% endif %}
    {% if alignfeatplot  %}
      <td><a href="{{alignfeatplot}}" title="A graphical summary of the distribution of reads to genomic features."><img style="height:auto" src="{{alignfeatplot}}" alt="Feature Alignment Distribution"/></a></td>
    {% endif %}
    </tr><tr>
    {% if transcovplot %}
      <td><a href="{{transcovplot}}" title="A overlay of individual normalized transcript coverage plots for each barcode."><img style="height:auto" src="{{transcovplot}}" alt="Normalized Transcript Coverage"/></a></td>
    {% endif %}
    {% if genepdfplot %}
      <td><a href="{{genepdfplot}}" title="Distribution of genes across barcodes showing the frequency of numbers of genes having similar log10 reads counts. All curves are plotted on the same axis scale. The counts data is fitted to a gaussian kernel using the default R 'density' function."><img style="height:auto" src="{{genepdfplot}}" alt="Distribution of Gene Reads"/></a></td>
    {% endif %}
    </tr><tr>
    {% if isofdistplot %}
      <td colspan="2" style="text-align:center"><a href="{{isofdistplot}}" title="Distribution of transcript isoforms across barcodes showing the counts of isoforms having similar FPKM values. All curves are plotted on the same y-axis, normalized to the highest count and scaled for FPKM values >= {{fpkm_thres}}."><img style="height:auto" src="{{isofdistplot}}" alt="Distribution of Isoform Reads"/></a></td>
    {% endif %}
    </tr></table>
  </div>
{% if rvalheatmap %}
  <div class="content" id="CorHeatmap">
    <div style="text-align:center">
      <a href="{{rvalheatmap}}" title="A heatmap of Spearman correlation r-values from comparing log2 RPM reads pair correlations barcodes, with dendrogram reflecting ordering of barcodes as being most similar by these values."><img style="height:auto" src="{{rvalheatmap}}" alt="Heatmap of Barcode Pair Correlations"/></a>
    </div>
  </div>
{% endif %}
{% if cpairsplot %}
  <div class="content" id="CorPlot">
    <div style="text-align:center">
      <a class="unscaled_image_window" href="{{cpairsplot}}" title='Barcode read pair correlation plot. Lower panels show log2(RPM+1) values plotted for each pair of barcodes, with linear least squares regression line overlaid and line slope reported. Upper panels show Pearson Correlation r-value for the regression line. Diagonal panels shows the frequency density plot for the individual log(RPM+1) values for each barcode. (If only a single barcode had reads then just the density plot is displayed.) Click to open an expanded view in a new window.'><img id="paircor_img" style="width:auto" src="{{cpairsplot}}" alt="Barcode read pair correlation plot."/></a>
    </div>
  </div>
{% endif %}
{% if geneheatmap %}
  <div class="content" id="GeneHeatmap">
    <div style="width:50%;margin:0 auto;width:800px;height:900px;border:1px solid #ccc;overflow:auto">
      <a href="{{geneheatmap}}" title="A heatmap of {{hm_samples|intcomma}} genes showing the most variation in representation across barcodes as measured by the coefficient of variation (CV) of normalized reads counts for genes that have at least one barcode with at least {{hm_minrpm|intcomma}} RPM reads, plotted using log10 of those counts. For this plot barcodes will be omitted if they had less than {{hm_bcmrds|intcomma}} total reads."><img src="{{geneheatmap}}" alt="Heatmap of Barcode log10(RPM) for {{hm_samples|intcomma}} genes showing largest variation across barcodes."/></a>
    </div>
  </div>
{% endif %}
{% if isofheatmap %}
  <div class="content" id="IsoformHeatmap">
    <div style="width:50%;margin:0 auto;width:1000px;height:900px;border:1px solid #ccc;overflow:auto">
      <a href="{{isofheatmap}}" title="A heatmap of up to {{hm_samples|intcomma}} gene transcript isoforms showing the most variation in representation across barcodes as measured by the coefficient of variation (CV) of FPKM values for isoforms that have an FKPM value >= {{hm_minfpkm}} for at least one barcode, plotted using log10 of FKPM+1. Barcodes are excluded if they had less than {{hm_bcmexp|intcomma}} isoforms detected at FPKM values >= {{fpkm_thres}}."><img src="{{isofheatmap}}" alt="Heatmap of Barcode log10(FPKM) for {{hm_samples|intcomma}} genes showing largest variation across barcodes."/></a>
    </div>
  </div>
{% endif %}
</div>
{% endif %}

{% if autorefresh %}
<br/><h3 style="text-align:center;color:red">*** Analysis is not complete ***</h3>
<a href="javascript:document.location.reload();" ONMOUSEOVER="window.status='Refresh'; return true">
<div style="text-align:center">Click here to refresh</div></a>
{% endif %}
{% if warning %}
<br/><h3 style="text-align:left;color:red">Warning: {{warning}}</h3>
{% endif %}

</div>

<h3 style='margin-left:-7px'>
{% if bctable %}
  <li><a href='{{bctable}}' title='Click to download a table file of the Barcode Summary Report presented above.'>Download Barcode Summary Report</a></li>
{% endif %}
{% if bcmatrix %}
<li><a href="{{bcmatrix}}" title='Click to download a table file of reads assigned to individual {{bcmtype}}s for each barcode.'>Download absolute reads table</a></li>
{% endif %}
{% if rpmbcmatrix %}
<li><a href="{{rpmbcmatrix}}" title='Click to download a table file of RPM (Reads Per Million reads) values assigned to individual {{bcmtype}}s for each barcode.'>Download absolute normalized reads table</a></li>
{% endif %}
{% if featmatrix %}
<li><a href="{{featmatrix}}" title='Click to download a table file of read counts assigned to genomic features, as used to create the Alignment Distribution plot.'>Download aligned reads distribution table</a></li>
{% endif %}
<!--
{% if rvalmatrix %}
<li><a href="{{rvalmatrix}}" title='Click to download a table file of log2(RPM) correlation r-values for pairs of barcodes, as used in the correlation and heatmap plots. This file is output as in R matrix format.'>Download log2(RPM) pair correlation r-value matrix</a></li>
{% endif %}
-->
{% if isofmatrix %}
<li><a href="{{isofmatrix}}" title='Click to download a table file of all gene isoform FPKM values assigned using cufflinks for all barcodes..'>Download isoform FPKM values table</a></li>
{% endif %}
</h3>

<div class="footer" style="width:100%"> <div class="container"> <div class="row"> <div class="span12">
  <img class="pull-left" src="/site_media/resources/img/ion-logo2.png" border="0">
  <div class="pull-right">
    <p>This is for research use only. It is not intended for any animal or human therapeutic or diagnostic use.</p>
    <p><span class="copyright">&copy; 2014 Life Technologies Corporation. All Rights Reserved.</span>

    <a href="/licenses/terms-of-use.txt">Terms &amp; Conditions</a></p>
  </div>
</div> </div> </div> </div>
<br/>
</body> 
</html>
