<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE HTML>
<html>
<base target="_parent"/>
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=utf-8">
<head>
<link type="text/css" rel="stylesheet" href="/site_media/resources/styles/tb-layout.css" />
<link type="text/css" rel="stylesheet" href="/site_media/resources/styles/tb-styles.min.css" />
<link type="text/css" rel="stylesheet" href="/site_media/jquery/css/aristo/jquery-ui-1.8.7.custom.css" />
<link type="text/css" rel="stylesheet" href="/site_media/resources/styles/print.css" media="print" />

<link type="text/css" rel="stylesheet" href="lifechart/pseudo-bootstrap.css" />
<link rel="stylesheet" type="text/css" href="lifechart/lifegrid.css"/>

{% load humanize %}

<style type="text/css">
  body {
    background:white;
    padding: 0;
    margin: 8px;
  }
  .block-rounded {
    padding:5px;
    background-color:#F5F5F5;
    border:1px solid #D6D6D6;
    -moz-border-radius: 6px;
    -webkit-border-radius:6px;
    border-radius: 6px
  }
  .help {
    cursor:help;
    border-bottom: 1px dotted #A9A9A9
  }
</style>

<title>RNASeq Analysis Report</title>

</head>

<body>

<h1><center>RNASeq Analysis Report</center></h1>
{% if Sample_Name != "" and Sample_Name != "None" %}
<h3><center>Sample Name: {{Sample_Name}}</center></h3>
{% endif %}
<h3><center>{{run_name}}<center></h3>

<table class="center">
<tr><td><h4>
  <b>Reference:</b> {{Reference_Genome}}<br/>
</h4></td></tr><tr><td>

<table class="center block-rounded"><tr>
  <td><a href="{{plot_generep}}" title="Simple plot showing the numbers of genes with reads in log10 counting bins."><img style="height:140px;margin-top:6px" src="{{plot_generep}}" alt="Gene Representation"/></a></td>
  <td><div class="statsdata">
  <table>
    <tr><td class="inleft"><span title="Number of unmapped reads">Total reads</span></td>
        <td class="inright">{{Total_Reads|intcomma}}</td></tr>
    <tr><td class="inleft"><span title="Number of aligned reads.">Aligned reads</span></td>
        <td class="inright">{{Aligned_Reads|intcomma}}</td></tr>
    <tr><td class="inleft"><span title="Percentage of total reads that were aligned to the reference (as transcripts).">Percent aligned reads</span></td>
        <td class="inright">{{Pct_Aligned}}</td></tr>
    <tr><td class="inleft"><span title="Mean average read length of aligned reads.">Mean read length</span></td>
        <td class="inright">{{Mean_Read_Length}}</td></tr>
    <tr><td class="inleft"><span title="Number of annotated genes with at least 10 reads.">Genes Detected</span></td>
        <td class="inright">{{Genes_with_10__reads|intcomma}}</td></tr>
    <tr><td class="inleft"><span title="Number of isoforms detected by cufflinks at FPKM >= {{fpkm_thres}}.">Isoforms Detected</span></td>
        <td class="inright">{{Isoforms_Detected|intcomma}}</td></tr>
  </table></div></td>
</tr></table>
</td></tr><tr><td>

<br/>
<table class="center block-rounded">
  <tr>
    <th><span style="cursor:help" title="Summary of reads mapped to genes of annotated reference.">Gene Mapping Summary</span></th>
    <th><span style="cursor:help" title="Summary of base reads aligned to genetic features of annotated reference.">Base Mapping Summary</span></th>
  </tr>
  <tr>
    <td><div class="statsdata"><table>
       <tr><td class="inleft"><span title="Number of genes in annotated reference.">Reference genes</span></td>
           <td class="inright">{{Reference_Genes|intcomma}}</td></tr>
       <tr><td class="inleft"><span title="Total number of reads aligned to reference genes.">Reads mapped to genes</span></td>
           <td class="inright">{{Reads_Mapped_to_Genes|intcomma}}</td></tr>
       <tr><td class="inleft"><span title="Genes that had at least 1 aligned read.">Genes with 1+ reads</span></td>
           <td class="inright">{{Genes_with_1__reads|intcomma}}</td></tr>
       <tr><td class="inleft"><span title="Genes that had at least 10 aligned reads.">Genes with 10+ reads</span></td>
           <td class="inright">{{Genes_with_10__reads|intcomma}}</td></tr>
       <tr><td class="inleft"><span title="Genes that had at least 100 aligned reads.">Genes with 100+ reads</span></td>
           <td class="inright">{{Genes_with_100__reads|intcomma}}</td></tr>
       <tr><td class="inleft"><span title="Genes that had at least 1,000 aligned reads.">Genes with 10,00+ reads</span></td>
           <td class="inright">{{Genes_with_1000__reads|intcomma}}</td></tr>
       <tr><td class="inleft"><span title="Genes that had at least 10,000 aligned reads.">Genes with 10,000+ reads</span></td>
           <td class="inright">{{Genes_with_10000__reads|intcomma}}</td></tr>
       <tr><td class="inleft"><span title="Number of gene transcript isoforms annotated.">Isoforms Annotated</span></td>
           <td class="inright">{{Isoforms_Annotated|intcomma}}</td></tr>
       <tr><td class="inleft"><span title="Number of isoforms detected by cufflinks at FPKM >= {{fpkm_thres}}.">Isoforms Detected</span></td>
           <td class="inright">{{Isoforms_Detected|intcomma}}</td></tr>
    </table></div></td>
    <td><div class="statsdata"><table>
      <tr><td class="inleft"><span title="Total bases in all reads">Total base reads</span></td>
          <td class="inright">{{Total_Base_Reads|intcomma}}</td></tr>
      <tr><td class="inleft"><span title="Total aligned bases in all mapped reads">Total aligned bases</span></td>
          <td class="inright">{{Total_Aligned_Bases|intcomma}}</td></tr>
      <tr><td class="inleft"><span title="Percentage of base reads that were aligned to the reference.">Percent aligned bases</span></td>
          <td class="inright">{{Pct_Aligned_Bases}}</td></tr>
      <tr><td class="inleft"><span title="Percentage of aligned bases in mRNA coding regions of the reference.">Percent coding bases</span></td>
          <td class="inright">{{Pct_Coding_Bases}}</td></tr>
      <tr><td class="inleft"><span title="Percentage of aligned bases in (mRNA) untranslated regions (UTRs) of the reference.">Percent UTR bases</span></td>
          <td class="inright">{{Pct_UTR_Bases}}</td></tr>
      <tr><td class="inleft"><span title="Percentage of aligned bases in ribosomal RNA regions of the reference.">Percent ribosomal bases</span></td>
          <td class="inright">{{Pct_Ribosomal_Bases}}</td></tr>
      <tr><td class="inleft"><span title="Percentage of aligned bases in intronic regions of the reference.">Percent intronic bases</span></td>
          <td class="inright">{{Pct_Intronic_Bases}}</td></tr>
      <tr><td class="inleft"><span title="Percentage of aligned bases in intergenic regions of the reference.">Percent intergenic bases</span></td>
          <td class="inright">{{Pct_Intergenic_Bases}}</td></tr>
      <tr><td class="inleft"><span title="Fraction of (forward) reads aligned to the sense reference strand to total aligned reads.">Strand balance</span></td>
          <td class="inright">{{Strand_Balance}}</td></tr>
    </table></div></td>
  </tr>
</table>
</td></tr><tr><td>

<table class="center"><tr>
{% if plot_transcov %}
  <td><a href="{{plot_transcov}}" title="A plot of normalized transcript coverage; the frequency of base reads with respect to the lengths of individual transcript they are aligned to, in the 3' to 5' orientation."><img style="width:350px" src="{{plot_transcov}}" alt="Normalized Transcript Coverage"/></a></td>
{% endif %}
{% if plot_rnareads %}
  <td><a href="{{plot_rnareads}}" title="A pie chart showing the relative distribution of (base) reads to genomic features."><img style="width:350px" src="{{plot_rnareads}}" alt="Feature Alignment Distribution"/></a></td>
{% endif %}
{% if plot_isoexpr %}
 </tr><tr>
  <td colspan="2" style="text-align:center"><a href="{{plot_isoexpr}}" title="Boxplots showing variation of isofroms expressed at FPKM >= {{fpkm_thres}} for each set of genes grouped by ther number anticipated (annotated) isoforms. Wiskers are defined by points within Q1-1.5xIQR to Q3+1.5xIQR. Only genes with 25 or less isoforms are represented in this plot. The data and a plot for all genes is available for download using the links at the bottom of this report."><img src="{{plot_isoexpr}}" alt="Expressed Isoforms per Gene"/></a></td>
 </tr><tr>
{% endif %}
</tr></table>

<br/>
<h3 style='margin-left:-7px;line-height:20px'>
  <li><a href='{{stats}}'>Download the Statistics Summary</a></li>
  <li><a href='{{genereads}}'>Download the Gene Read Counts</a></li>
  <li><a href='{{dl_files}}'>Download Output Files (page)</a></li>
  <li><a href='{{cufflinks}}'>Download Cufflinks Output Files (page)</a></li>
</h3>
</td></tr>
</table>

<br/>
<div class="footer" style="width:100%"> <div class="container"> <div class="row"> <div class="span12">
  <img class="pull-left" src="/site_media/resources/img/ion-logo2.png" border="0">
  <div class="pull-right">
    <p>This is for research use only. It is not intended for any animal or human therapeutic or diagnostic use.</p>
    <p><span class="copyright">&copy; 2014 Life Technologies Corporation. All Rights Reserved.</span>

    <a href="/licenses/terms-of-use.txt">Terms &amp; Conditions</a></p>
  </div>
</div> </div> </div> </div>
<br/>
</body>
</html>

